internal/client/ssh.go

```diff
@@ -0,0 +1,386 @@
+	Password              string
+	PKPath                string
+	InsecureIgnoreHostKey bool
+
+	conn        *ssh.Client
+	connHasRoot bool
+}
+
+func (c *SSHClient) newSSHSession() (*ssh.Session, error) {
+	if c == nil || c.conn == nil {
+		return nil, fmt.Errorf("not connected")
+	}
+
+	return c.conn.NewSession()
+}
+
+func (c *SSHClient) hasRootPrivileges() (bool, error) {
+	session, err := c.newSSHSession()
+	if err != nil {
+		return false, err
+	}
+	defer session.Close()
+
+	output, err := session.Output("whoami")
+	if err != nil {
+		return false, fmt.Errorf("failed to determine if root: (%s, %s)", err.Error(), string(output))
+	}
+
+	c.connHasRoot = (string(output) == "root")
+	return c.connHasRoot, nil
+}
+
+func (c *SSHClient) isServiceInstalled(service string) (bool, error) {
+	session, err := c.newSSHSession()
+	if err != nil {
+		return false, err
+	}
+	defer session.Close()
+
+	output, err := session.Output(fmt.Sprintf("systemctl list-unit-files %s", service))
+	if err != nil {
+		if serr, ok := err.(*ssh.ExitError); ok && serr.ExitStatus() == 1 {
+			return false, nil
+		}
+		return false, fmt.Errorf("failed to list unit files: %s", string(output))
+	}
+
+	return strings.Contains(string(output), service), nil
+}
+
+func (c *SSHClient) determineOS() (string, error) {
+	session, err := c.newSSHSession()
+	if err != nil {
+		return "", err
+	}
+	defer session.Close()
+
+	output, err := session.Output("cat /etc/os-release | grep -w ID=")
+	if err != nil {
+		return "", fmt.Errorf("failed to determine os: %s", string(output))
+	}
+
+	os := strings.ReplaceAll(string(output), "ID=", "")
+	os = strings.ReplaceAll(os, "\"", "")
+	return strings.TrimSpace(os), nil
+}
+
+func (c *SSHClient) Connect() error {
+	var err error
+	var authMethods []ssh.AuthMethod
+	var hostkeyCallback ssh.HostKeyCallback
+
+	if c.PKPath != "" {
+		pkPath := c.PKPath
+		// resolve home directory
+		if strings.HasPrefix(pkPath, "~/") {
+			homeDir, err := os.UserHomeDir()
+			if err != nil {
+				return err
+			}
+			pkPath = strings.Replace(pkPath, "~/", homeDir+"/", 1)
+		}
+		pKey, err := os.ReadFile(pkPath)
+		if err != nil {
+			return err
+		}
+
+		signer, err := ssh.ParsePrivateKey(pKey)
+		if err != nil {
+			return err
+		}
+
+		authMethods = append(authMethods, ssh.PublicKeys(signer))
+	}
+
+	if c.Password != "" {
+		authMethods = append(authMethods, ssh.Password(c.Password))
+	}
+
+	hostkeyCallback = ignoreHostKeyCallback
+	if !c.InsecureIgnoreHostKey {
+		homeDir, err := os.UserHomeDir()
+		if err != nil {
+			return err
+		}
+
+		knownHostsPath := fmt.Sprintf("%s/.ssh/known_hosts", homeDir)
+		hostkeyCallback, err = knownhosts.New(knownHostsPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	conf := &ssh.ClientConfig{
+		User:            c.User,
+		HostKeyCallback: hostkeyCallback,
+		Auth:            authMethods,
+	}
+
+	c.conn, err = ssh.Dial("tcp", c.Host, conf)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.hasRootPrivileges()
```

Comment:
`hasRootPrivileges` is an internal method used only here. If you don't need its return value, don't return it

Reply:
Good point. Changed.